draft-irtf-cfrg-vdaf.md

```diff
@@ -1040,46 +1040,62 @@ DAFs MUST implement the following function:
 
-    out_share_0_0   out_share_1_0       out_share_[SHARES-1]_0
-    out_share_0_1   out_share_1_1       out_share_[SHARES-1]_1
-    out_share_0_2   out_share_1_2       out_share_[SHARES-1]_2
-         ...             ...                     ...
-    out_share_0_B   out_share_1_B       out_share_[SHARES-1]_B
-      |               |                   |
-      V               V                   V
-    +-----------+   +-----------+       +-----------+
-    | aggregate |   | aggregate |   ... | aggregate |
-    +-----------+   +-----------+       +-----------+
-      |               |                   |
-      V               V                   V
-    agg_share_0     agg_share_1         agg_share_[SHARES-1]
 ~~~~
-{: #aggregate-flow title="Aggregation of output shares. `B` indicates the number of
-measurements in the batch."}
+    Aggregator j
+    ============
+
+    +------------+    +------------+
+    | agg_init   |--->| agg_update |<--- out_share_0
+    +------------+    +------------+
+                         |
+                         V
+                      +------------+
+                      | agg_update |<--- out_share_1,
+                      +------------+
+                        |
+                        V
+                       ...
+                        |
+                        V
+                      +------------+
+                      | agg_update |<--- out_share_MEAS_COUNT
+                      +------------+
+                        |
+                        V
+                      agg_share_j
+~~~~
+{: #aggregate-flow title="Local aggregation of output shares by Aggregator j.
+The number of measurements in the batch is denoted by MEAS_COUNT."}
+
+In many situations it is desirable to split an aggregate share across multiple
+storage elements, then merge the aggregate shares together just before
+releasing the completed aggregate share to the Collector. DAFs facilitate this
+with the following method:
+
+* `daf.merge(agg_param: AggParam, agg_shares: list[AggShare]) -> AggShare`
+  merges a sequence of aggregate shares into a single aggregate share.
 
-For simplicity, we have written this algorithm in a "one-shot" form, where all
-output shares for a batch are provided at the same time. Many DAFs may also
-support a "streaming" form, where shares are processed one at a time.
+### Aggregation Order {#agg-order}
 
-Implementation note: for most natural DAFs (and VDAFs) it is not necessary for
-an Aggregator to store all output shares individually before aggregating.
-Typically it is possible to merge output shares into aggregate shares as they
-arrive, merge these into other aggregate shares, and so on. In particular, this
-is the case when the output shares are vectors over some finite field and
-aggregating them involves merely adding up the vectors element-wise. Such is the
-case for Prio3 {{prio3}} and Poplar1 {{poplar1}}.
+For most DAFs and VDAFs, the outcome of aggregation is not sensitive to the
```

Comment:
I forgot about this wrinkle; I don't think we mention in DAP that we can only work with VDAFs that allow aggregating output shares in an arbitrary order.

Do you think it is important to maintain that VDAFs might require a specific aggregation ordering? If not, let's drop this section (and maybe replace it with a note that aggregation can occur in any order). If so, I'll update DAP's VDAF requirements to note the arbitrary-ordering requirement.

Reply:
Yup, this is indeed one of our requirements for DAP.

It seems unlikely, but the possibility has been brought up by enough people that I don't think we should rule it out completely. Better I think to document the requirement as part off https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/581.

Reply:
Closing the loop: I sent https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/603 for this.